### ISSUE TITLE
Xml and XmlItem are printed differently

### ISSUE THREAD

**Issue by elordin** [links-lang/links#256]
When running links in web-mode:
```
links> hd(<div id="myId" class="a-class" />);
<div id="myId"class="a-class"/> : XmlItem
```
```
links> <div id="myId" class="a-class" />;
<div id="myId" class="a-class" /> : Xml
```
Clearly the `XmlItem` version should have a space between attributes as well.

**Comment by jamescheney** [comment]
Also:

```
links> hd(<foo a="1" b="2"><bar a="3" b="4"><baz a="5" b="6"></baz></bar><bar a="42" b="17"></bar></foo>);
<foo b="2" a="1"><bar b="4" a="3"><baz b="6" a="5"/></bar><bar b="17"a="42"/></foo> : XmlItem
```

That is, things work right for `XmlItem`s that have content, hence their close tags must be printed, but wrong for `XmlItem`s that have attributes but no element content.

So perhaps the bug is conditionally guarded somewhere by a `close_tags` test.

**Comment by jamescheney** [comment]
I suggest we defer this to 0.7.3 and release 0.7.2 ASAP.

**Comment by jamescheney** [comment]
I think fixing this should be a part of the general list/xml cleanup that will hopefully fall out naturally if we make lists and XML into user-defined types #46, so happy to kick to >0.7.3.

**Comment by jamescheney** [comment]
I've looked into this and it seems to be due to some weird stateful behavior regarding how `pp_print_space` hints are treated by the `Format` library.  For some reason if an `Xml = [XmlItem]` is printed by printing a list using the general value printer the right thing happens, but if the `p_xmlitem` printer is called directly on a singleton `XmlItem` then certain space hints are not printed as spaces.  

The reason spaces seem to be being dropped between attributes is that zero-width space hints are being generated, and the code in `Value.string_of_pretty` is setting the buffer formatter's `out_spaces` function to print zero spaces if there is a zero-width hint, or one space (instead of many) otherwise.  This code is also only being called because the `print_pretty` setting tested in  `Repl.print_value` is false.  If we set `print_pretty` to true, the problem seems to go away, but probably we still want XML to be printed correctly even if pretty-printing is disabled.  (And I'm not sure why pretty-printing is disabled by default in the first place.)
